c10/core/impl/cow/COW.cpp

```diff
@@ -30,24 +29,18 @@ at::DataPtr copy_data_ptr(at::DataPtr const& data_ptr) {
   return make_data_ptr(data_ptr, *ctx);
 }
 
-bool is_simple_context(
-    const void* context,
-    const void* data,
-    const at::Allocator* allocator) {
-  if (allocator == c10::GetDefaultMobileCPUAllocator()) {
```

Comment:
I realized that I could remove the dependence on CPUAllocator here by adding a new method `Allocator::is_simple_data_ptr`. It's better for CPUAllocator to override that method than to have a conditional here in `COW.cpp`

Also, since `StorageImpl` now needs to have access to `materialize_cow_storage`, I needed to combine the `impl_cow` bazel library back into the `base` library